local/php/pie.go

```diff
@@ -0,0 +1,141 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package php
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/symfony-cli/symfony-cli/util"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type PieResult struct {
+	code  int
+	error error
+}
+
+func (p PieResult) Error() string {
+	if p.error != nil {
+		return p.error.Error()
+	}
+
+	return ""
+}
+
+func (p PieResult) ExitCode() int {
+	return p.code
+}
+
+func Pie(dir string, args, env []string, stdout, stderr, logger io.Writer, debugLogger zerolog.Logger) PieResult {
+	e := &Executor{
+		Dir:        dir,
+		BinName:    "php",
+		Stdout:     stdout,
+		Stderr:     stderr,
+		SkipNbArgs: -1,
+		ExtraEnv:   env,
+		Logger:     debugLogger,
+	}
+
+	if piePath := os.Getenv("SYMFONY_PIE_PATH"); piePath != "" {
+		debugLogger.Debug().Str("SYMFONY_PIE_PATH", piePath).Msg("SYMFONY_PIE_PATH has been defined. User is taking control over PIE detection and execution.")
+		e.Args = append([]string{piePath}, args...)
+	} else if path, err := e.findPie(); err == nil && assertIsPHPScript(path) {
+		e.Args = append([]string{"php", path}, args...)
+	} else {
+		reason := "No PIE installation found."
+		if path != "" {
+			reason = fmt.Sprintf("Detected PIE file (%s) is not a valid PHAR or PHP script.", path)
+		}
+		fmt.Fprintln(logger, "  WARNING:", reason)
+		fmt.Fprintln(logger, "  Downloading PIE for you, but it is recommended to install PIE yourself, instructions available at https://github.com/php/pie")
+		// we don't store it under bin/ to avoid it being found by findPie as we want to only use it as a fallback
+		binDir := filepath.Join(util.GetHomeDir(), "pie")
+		if path, err = downloadPie(binDir); err != nil {
+			return PieResult{
+				code:  1,
+				error: errors.Wrap(err, "unable to find pie, get it at https://github.com/php/pie"),
+			}
+		}
+		e.Args = append([]string{"php", path}, args...)
+		fmt.Fprintf(logger, "  (running %s)\n\n", e.CommandLine())
+	}
+
+	ret := e.Execute(false)
+	if ret != 0 {
+		return PieResult{
+			code:  ret,
+			error: errors.Errorf("unable to run %s", e.CommandLine()),
+		}
+	}
+	return PieResult{}
+}
+
+func findPie(logger zerolog.Logger) (string, error) {
+	for _, file := range []string{"pie", "pie.phar"} {
+		logger.Debug().Str("source", "PIE").Msgf(`Looking for PIE in the PATH as "%s"`, file)
+		if pharPath, _ := LookPath(file); pharPath != "" {
+			// On Windows, we don't want the .bat, but the real pie phar/PHP file
+			if strings.HasSuffix(pharPath, ".bat") {
+				pharPath = pharPath[:len(pharPath)-4] + ".phar"
+			}
```

Comment:
I’m not familiar with PIE but I don’t recall a bat wrapper being shipped with PIE.
Do you have more info about this aspect?

Reply:
You're right, it's an artifact from the Composer wrapper. I removed it